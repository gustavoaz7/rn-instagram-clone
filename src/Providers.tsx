import React, { ReactNode } from 'react';
import { ThemeProvider } from 'styled-components/native';
import { Provider, ProviderProps } from 'react-redux';
import { PersistGate } from 'redux-persist/integration/react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { RootSiblingParent } from 'react-native-root-siblings'; // dep of rn-root-toast
import { theme as stylesTheme } from './styles/theme';
import { persistor, store as reduxStore } from './redux/store';
import { useThemeVariantSelector } from './redux/theme-variant';

type TProvidersProps = {
  children?: ReactNode;
  store?: ProviderProps['store'];
  theme?: typeof stylesTheme;
};

export function Providers({
  children = null,
  store = reduxStore,
  theme = stylesTheme,
}: TProvidersProps): JSX.Element {
  return (
    <RootSiblingParent>
      <Provider store={store}>
        <PersistGate loading={null} persistor={persistor}>
          <Theming theme={theme}>{children}</Theming>
        </PersistGate>
      </Provider>
    </RootSiblingParent>
  );
}

function Theming({
  theme,
  children,
}: Required<Omit<TProvidersProps, 'store'>>): JSX.Element {
  const themeVariant = useThemeVariantSelector();
  return <ThemeProvider theme={theme[themeVariant]}>{children}</ThemeProvider>;
}
