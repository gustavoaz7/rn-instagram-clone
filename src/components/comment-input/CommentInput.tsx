import React, { memo, useCallback, useEffect, useRef, useState } from 'react';
import { Animated, StyleSheet } from 'react-native';
import styled, { useTheme } from 'styled-components/native';
import { EMOJIS } from '../../constants';
import { useUserSelector } from '../../redux/user';
import { SCREEN_WIDTH } from '../../utils/dimensions';
import { AvatarWithRing } from '../avatar-with-ring';
import { Text } from '../text';

export type TCommentInputProps = {
  onSubmit: (text: string) => void;
};

export const CommentInput = ({ onSubmit }: TCommentInputProps): JSX.Element => {
  const { user } = useUserSelector();
  const [text, setText] = useState('');
  const handleTextChange = useCallback(newText => {
    setText(newText);
  }, []);
  const handleEmojiPress = useCallback(
    emoji => {
      setText(`${text}${emoji}`);
    },
    [text],
  );
  const handlePostPress = useCallback(() => {
    onSubmit(text);
  }, [onSubmit, text]);

  return (
    <Container>
      <EmojisRow>
        {EMOJIS.map(emoji => (
          <CommentInput.Emoji
            key={emoji}
            emoji={emoji}
            onPress={handleEmojiPress}
          />
        ))}
      </EmojisRow>
      <InputRow>
        <AvatarWithRing
          size={40}
          color="transparent"
          imageUrl={user?.profilePicUrl}
        />
        <Input
          value={text}
          onChangeText={handleTextChange}
          placeholder="Add a comment..."
          multiline
        />
        <PostButton disabled={!text.length} onPress={handlePostPress}>
          <PostText>Post</PostText>
        </PostButton>
      </InputRow>
    </Container>
  );
};

type TCommentInputEmojiProps = {
  emoji: string;
  onPress: (text: string) => void;
};
CommentInput.Emoji = memo(
  ({ emoji, onPress }: TCommentInputEmojiProps): JSX.Element => {
    const theme = useTheme();
    const [isPressing, setIsPressing] = useState(false);
    const scale = useRef(new Animated.Value(1)).current;

    useEffect(() => {
      Animated.timing(scale, {
        toValue: isPressing ? 0.9 : 1,
        duration: theme.animation.timingFast,
        useNativeDriver: true,
      }).start();
    }, [isPressing, scale, theme.animation.timingFast]);

    return (
      <Animated.View style={{ transform: [{ scale }] }}>
        <EmojiContainer
          onPress={() => onPress(emoji)}
          onPressIn={() => setIsPressing(true)}
          onPressOut={() => setIsPressing(false)}
        >
          <Emoji>{emoji}</Emoji>
        </EmojiContainer>
      </Animated.View>
    );
  },
);

const Container = styled.View``;

const Row = styled.View`
  flex-direction: row;
`;

const EmojisRow = styled(Row)`
  align-items: center;
  border-color: ${({ theme }) => theme.color.gray};
  border-top-width: ${StyleSheet.hairlineWidth}px;
  border-bottom-width: ${StyleSheet.hairlineWidth}px;
`;

const InputRow = styled(Row)`
  padding: ${({ theme }) => `${theme.spacing.xs} ${theme.spacing.m}`};
  align-items: baseline;
`;

const EmojiContainer = styled.Pressable`
  width: ${SCREEN_WIDTH / EMOJIS.length}px;
  height: ${SCREEN_WIDTH / EMOJIS.length}px;
  justify-content: center;
  align-items: center;
`;

const Emoji = styled(Text)`
  font-size: ${SCREEN_WIDTH / (EMOJIS.length * 2)}px;
`;

const Input = styled.TextInput`
  flex: 1;
  padding: 0 ${({ theme }) => theme.spacing.m};
  max-height: 120px;
`;

const PostButton = styled.Pressable`
  opacity: ${({ disabled }) => (disabled ? 0.5 : 1)};
`;

const PostText = styled(Text)`
  color: ${({ theme }) => theme.color.lightBlue};
  padding-bottom: ${({ theme }) => theme.spacing.xs};
`;
