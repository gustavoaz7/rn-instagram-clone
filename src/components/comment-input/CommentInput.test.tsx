import React from 'react';
import { TextInput } from 'react-native';
import {
  render,
  RenderOptions,
  fireEvent,
} from '@testing-library/react-native';
import { CommentInput, EMOJIS } from './CommentInput';
import { Providers } from '../../Providers';

describe('components - CommentInput', () => {
  const submitSpy = jest.fn();
  const props = { onSubmit: submitSpy };
  const options: RenderOptions = { wrapper: Providers };

  it('renders', () => {
    render(<CommentInput {...props} />, options);
  });

  it('matches snapshot', () => {
    const { toJSON } = render(<CommentInput {...props} />, options);

    expect(toJSON()).toMatchSnapshot();
  });

  it('renders emojis list', () => {
    const { getByText } = render(<CommentInput {...props} />, options);

    EMOJIS.forEach(emoji => {
      expect(getByText(emoji)).toBeTruthy();
    });
  });

  describe('when post button is pressed', () => {
    describe('when text input is not empty', () => {
      it('does not call onSubmit', () => {
        const text = `My number: ${Math.random()}`;
        const { UNSAFE_getByType, getByText } = render(
          <CommentInput {...props} />,
          options,
        );

        const input = UNSAFE_getByType(TextInput);
        fireEvent.changeText(input, text);
        fireEvent.press(getByText('Post'));

        expect(submitSpy).toHaveBeenCalledTimes(1);
        expect(submitSpy).toHaveBeenCalledWith(text);
      });
    });

    describe('when text input is empty', () => {
      it('does not call onSubmit', () => {
        const { getByText } = render(<CommentInput {...props} />, options);

        fireEvent.press(getByText('Post'));

        expect(submitSpy).not.toHaveBeenCalled();
      });
    });
  });

  describe('when an emoji is pressed', () => {
    it('appends emoji to text input', () => {
      const prevText = 'My emoji: ';
      const { UNSAFE_getByType, getByText } = render(
        <CommentInput {...props} />,
        options,
      );

      const input = UNSAFE_getByType(TextInput);
      fireEvent.changeText(input, prevText);
      EMOJIS.forEach(emoji => {
        fireEvent.press(getByText(emoji));
      });

      expect(input.props.value).toBe(`${prevText}${EMOJIS.join('')}`);
    });
  });
});
