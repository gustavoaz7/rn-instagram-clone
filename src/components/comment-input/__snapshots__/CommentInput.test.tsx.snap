// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components - CommentInput matches snapshot 1`] = `
<View
  style={
    Array [
      Object {},
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderBottomWidth": 0.5,
          "borderColor": "#8e8e8e",
          "borderTopWidth": 0.5,
          "flexDirection": "row",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": 93.75,
              "justifyContent": "center",
              "width": 93.75,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#020202",
              },
              Object {
                "fontSize": 46.875,
              },
            ]
          }
        >
          ❤️
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": 93.75,
              "justifyContent": "center",
              "width": 93.75,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#020202",
              },
              Object {
                "fontSize": 46.875,
              },
            ]
          }
        >
          🙌
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": 93.75,
              "justifyContent": "center",
              "width": 93.75,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#020202",
              },
              Object {
                "fontSize": 46.875,
              },
            ]
          }
        >
          🔥
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": 93.75,
              "justifyContent": "center",
              "width": 93.75,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#020202",
              },
              Object {
                "fontSize": 46.875,
              },
            ]
          }
        >
          👏
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": 93.75,
              "justifyContent": "center",
              "width": 93.75,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#020202",
              },
              Object {
                "fontSize": 46.875,
              },
            ]
          }
        >
          😢
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": 93.75,
              "justifyContent": "center",
              "width": 93.75,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#020202",
              },
              Object {
                "fontSize": 46.875,
              },
            ]
          }
        >
          😍
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": 93.75,
              "justifyContent": "center",
              "width": 93.75,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#020202",
              },
              Object {
                "fontSize": 46.875,
              },
            ]
          }
        >
          😮
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "transform": Array [
            Object {
              "scale": 1,
            },
          ],
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": 93.75,
              "justifyContent": "center",
              "width": 93.75,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#020202",
              },
              Object {
                "fontSize": 46.875,
              },
            ]
          }
        >
          😂
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "baseline",
          "flexDirection": "row",
          "paddingBottom": 5,
          "paddingLeft": 15,
          "paddingRight": 15,
          "paddingTop": 5,
        },
      ]
    }
  >
    <View
      size={40}
      style={
        Array [
          Object {
            "borderRadius": 40,
            "height": 40,
            "overflow": "hidden",
            "position": "relative",
            "width": 40,
          },
        ]
      }
    >
      <ViewManagerAdapter_ExpoLinearGradient
        proxiedProperties={
          Object {
            "colors": Array [
              0,
              0,
            ],
            "endPoint": Array [
              0.3,
              0.7,
            ],
            "locations": undefined,
            "startPoint": Array [
              0.7,
              0.3,
            ],
          }
        }
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": "100%",
              "justifyContent": "center",
              "width": "100%",
            },
          ]
        }
      >
        <View
          ringWidth={2}
          size={40}
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#FFFFFF",
                "borderRadius": 36,
                "height": 36,
                "justifyContent": "center",
                "width": 36,
              },
            ]
          }
        >
          <Image
            offset={2}
            ringWidth={2}
            size={40}
            source={
              Object {
                "uri": undefined,
              }
            }
            style={
              Array [
                Object {
                  "borderColor": "#02020233",
                  "borderRadius": 32,
                  "borderWidth": 2,
                  "height": 32,
                  "width": 32,
                },
              ]
            }
          />
        </View>
      </ViewManagerAdapter_ExpoLinearGradient>
    </View>
    <TextInput
      allowFontScaling={true}
      multiline={true}
      onChangeText={[Function]}
      placeholder="Add a comment..."
      placeholderTextColor="#020202"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "color": "#020202",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "maxHeight": 120,
            "paddingBottom": 0,
            "paddingLeft": 15,
            "paddingRight": 15,
            "paddingTop": 0,
          },
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessible={true}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "opacity": 0.5,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#020202",
            },
            Object {
              "color": "#0095f6",
              "paddingBottom": 5,
            },
          ]
        }
      >
        Post
      </Text>
    </View>
  </View>
</View>
`;
